daml-lf/spec/daml-lf-1.rst

```diff
@@ -2493,48 +2375,258 @@ This section lists the built-in functions supported by DAML LF 1.
+`comparable` values. Comparable values are LF values except type
+abstractions, functions, partially applied builtin functions, and
+updates.
 
-  Returns ``'True'`` if the two argument are equal according ``~ᵥ``,
-  ``'False'`` otherwise.
+* ``LESS_EQ : ∀ (α:*). α → α → 'Bool'``
+
+  The builtin function ``LESS_EQ`` returns ``'True'`` if the first
+  argument is smaller than or equal to the second argument,
+  ``'False'`` otherwise. The function raises a runtime error if the
+  arguments are incomparable.
 
   [*Available in version >= 1.dev*]
 
-* ``LESS : ∀ (α:*). α → α → 'Bool'``
+  Formally the builtin function ``LESS_EQ`` semantics is defined by
+  the following rules. Note the rules assume ``LESS_EQ`` is fully
+  applied and well-typed, in particular ``LESS_EQ`` always compared
+  value of the same type.::
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqUnit
+      𝕆('LESS_EQ' @σ () ()) = Ok 'True'
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqBool
+      𝕆('LESS_EQ' @σ b₁ b₂) = Ok (¬b₁ ∨ b₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqInt64
+      𝕆('LESS_EQ' @σ LitInt64₁ LitInt64₂) = Ok (LitInt64₁ ≤ₗ LitInt64₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqDate
+      𝕆('LESS_EQ' @σ LitDate₁ LitDate₂) = Ok (LitDate₁ ≤ₗ LitDate₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqTimestamp
+      𝕆('LESS_EQ' @σ LitTimestamp₁ LitTimestamp₂) =
+          Ok (LitTimestamp₁ ≤ LitTimestamp₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqText
+      𝕆('LESS_EQ' @σ LitText₁ LitText₂) = Ok (LitText₁ ≤ₗ LitText₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqParty
+      𝕆('LESS_EQ' @σ LitParty₁ LitParty₂) = Ok (LitParty₁ ≤ₗ LitParty₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqNumeric
+      𝕆('LESS_EQ' @σ LitNumeric₁ LitNumeric₂) =
+          Ok (LitNumeric₁ ≤ LitNumeric₂)
+
+    —————————————————————————————————————————————————————————————————————— EvLessEqStructEmpty
+      𝕆('LESS_EQ' @⟨ ⟩ ⟨ ⟩ ⟨ ⟩) = Ok 'True'
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Err t
+    —————————————————————————————————————————————————————————————————————— EvLessEqStructNonEmptyHeadErr1
+      𝕆('LESS_EQ' @⟨ f₀: τ₁,  f₁: τ₁, …,  fₙ: τₙ ⟩
+                   ⟨ f₀= v₀,  f₁= v₁, …,  fₘ= vₘ ⟩
+   	           ⟨ f₀= v₀', f₁= v₁', …, fₘ= vₘ' ⟩) = Err t
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'False'
+    ————————————————————————————————————————————————————————————————————— EvLessEqStructNonEmptyHeadBigger
+      𝕆('LESS_EQ' @⟨ f₀: τ₁,  f₁: τ₁, …,  fₙ: τₙ  ⟩
+                   ⟨ f₀= v₀,  f₁= v₁, …,  fₘ= vₘ  ⟩
+      	           ⟨ f₀= v₀', f₁= v₁', …, fₘ= vₘ' ⟩) = Ok 'False'
+
+      𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'True'
+      𝕆('LESS_EQ' @τ₁ v₀' v₀) = Err t
```

Comment:
Do you anticipate this situation being possible? I.e. v0 <= v0' is fine but v0' <= v0 leads to error?

Reply:
I don't. I though it was simpler to specify explicitly this error case, than to prove this never happen.  This would indeed go away with the `compare` function way. I will try once last time to go this way (I abandon twice already).

Reply:
IMO it's fine if you want to keep the boolean presentation, but change this rule into a note that says that

```
     𝕆('LESS_EQ' @τ₁ v₀ v₀') = Ok 'True'
      𝕆('LESS_EQ' @τ₁ v₀' v₀) = Err t
```

cannot happen, and that if it happens, that's a bug.

Reply:
I simply drop  the rule. I gave up a third time trying to specify with the compare approach.